python/python/lancedb/table.py

```diff
@@ -117,15 +117,59 @@ def _sanitize_data(
         data = _sanitize_schema(
             data, schema=schema, on_bad_vectors=on_bad_vectors, fill_value=fill_value
         )
+        if schema is None:
+            schema = data.schema
     elif isinstance(data, Iterable):
         data = _to_record_batch_generator(
             data, schema, metadata, on_bad_vectors, fill_value
         )
+        if schema is None:
+            data, schema = _generator_to_data_and_schema(data)
+            if schema is None:
+                raise ValueError("Cannot infer schema from generator data")
     else:
         raise TypeError(f"Unsupported data type: {type(data)}")
     return data, schema
 
 
+def sanitize_create_table(
+    data, schema, metadata=None, on_bad_vectors="error", fill_value=0.0
+):
+    if inspect.isclass(schema) and issubclass(schema, LanceModel):
+        # convert LanceModel to pyarrow schema
+        # note that it's possible this contains
+        # embedding function metadata already
+        schema = schema.to_arrow_schema()
+
+    if data is not None:
+        data, schema = _sanitize_data(
+            data,
+            schema,
+            metadata=metadata,
+            on_bad_vectors=on_bad_vectors,
+            fill_value=fill_value,
+        )
+    else:
+        if schema is None:
+            raise ValueError("Either data or schema must be provided")
+        elif hasattr(data, "schema"):
+            schema = data.schema
+        elif isinstance(data, Iterable):
+            if metadata:
+                raise TypeError(
+                    (
+                        "Persistent embedding functions not yet "
+                        "supported for generator data input"
+                    )
+                )
+        data = pa.Table.from_pylist([], schema=schema)
```

Comment:
small nit: it's faster to not go through the Python -> Arrow conversion methods if you just want to create an empty table.

```suggestion
        data = pa.Table.from_batches([], schema=schema)
```

```python
In [5]: %timeit pa.Table.from_pylist([], schema)
6.1 µs ± 519 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [6]: %timeit pa.Table.from_batches([], schema)
287 ns ± 4.87 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

Reply:
Fixed